src/main/surfrdUtilsMod.F90

```diff
@@ -145,6 +146,86 @@ subroutine convert_cft_to_pft( begg, endg, cftsize, wt_cft )
+    ! Keep landunits above the user-defined thresholds and remove the rest
+    !
+    ! !USES:
+    use landunit_varcon, only: max_lunit, istsoil, istcrop, istice_mec, &
+                               istdlak, istwet, isturb_tbd, isturb_hd, &
+                               isturb_md
+    !
+    ! !ARGUMENTS:
+    integer, intent(in) :: begg  ! Beginning grid cell index
+    integer, intent(in) :: endg  ! Ending grid cell index
+    integer, intent(in) :: toosmall_soil  ! Soil landunit threshold (%)
+    integer, intent(in) :: toosmall_crop  ! Crop landunit threshold (%)
+    integer, intent(in) :: toosmall_glacier  ! Glacier landunit threshold (%)
+    integer, intent(in) :: toosmall_lake  ! Lake landunit threshold (%)
+    integer, intent(in) :: toosmall_wetland  ! Wetland landunit threshold (%)
+    integer, intent(in) :: toosmall_urb_tbd  ! Urban TBD landunit threshold (%)
+    integer, intent(in) :: toosmall_urb_hd  ! Urban HD landunit threshold (%)
+    integer, intent(in) :: toosmall_urb_md  ! Urban MD landunit threshold (%)
+    ! This array modified in-place
+    ! Weights of landunits per grid cell
+    real(r8), intent(inout) :: wt_lunit(begg:endg, max_lunit)
+    !
+    ! !LOCAL VARIABLES:
+    integer :: g  ! grid cell indexes
+    integer :: m  ! landunit indexes
+    integer :: max_landunit  ! landunit with largest fraction
+    integer :: toosmall_any  ! sum of the landunit thresholds
+    real(r8) :: toosmall(max_lunit)  ! Array of the thresholds (fraction)
+    real(r8) :: residual(max_lunit)  ! Array of wt_lunit residuals (fraction)
+    !-----------------------------------------------------------------------
+
+    SHR_ASSERT_ALL((ubound(wt_lunit) == (/endg, max_lunit/)), errMsg(sourcefile, __LINE__))
+
+    toosmall_any = toosmall_soil + toosmall_crop + toosmall_glacier + &
+                   toosmall_lake + toosmall_wetland + toosmall_urb_tbd + &
+                   toosmall_urb_hd + toosmall_urb_md
+
+    if (toosmall_any > 0) then
+
+       ! Copy the user-defined percent thresholds into array of fractions
+       toosmall(istsoil) = toosmall_soil / 100._r8
+       toosmall(istcrop) = toosmall_crop / 100._r8
+       toosmall(istice_mec) = toosmall_glacier / 100._r8
+       toosmall(istdlak) = toosmall_lake / 100._r8
+       toosmall(istwet) = toosmall_wetland / 100._r8
+       toosmall(isturb_tbd) = toosmall_urb_tbd / 100._r8
+       toosmall(isturb_hd) = toosmall_urb_hd / 100._r8
+       toosmall(isturb_md) = toosmall_urb_md / 100._r8
+
+       ! Loop through gridcells and landunits
+       do g = begg, endg
+          residual = 0._r8  ! initialize
+          do m = 1, max_lunit
+             ! Remove landunits that are too small
+             if (wt_lunit(g,m) > 0._r8 .and. wt_lunit(g,m) <= toosmall(m)) then
+                residual(m) = wt_lunit(g,m)
+                wt_lunit(g,m) = 0._r8
+             end if
+          end do
+          ! If all landunits got removed, go back and keep the largest landunit
```

Comment:
OK, this might need to be documented better. I think it makes sense, you have to fill the grid cell with at least one landunit, so it should be the biggest one. The user might want to be aware that could be the case though. I'm also wondering if this should have a warning if it's triggered, but I'm not sure.

Along the same lines I was wondering if you should be able to set one of the toosmalls to something over 100%, which would essentially trigger "I don't want to have ANY of the given landunit anywhere on my grid". If there's a desire to treat it that way, you would want different behavior in this case.

Reply:
@ekluzek I like the idea of a warning, though a user could end up with very many warnings if this happened in a large number of grid cells. I will include this warning:
`if ( masterproc ) then`
`write(iulog,*) 'WARNING: The values of namelist parameters toosmall_* have resulted in the removal of all landunits in grid cell g = ', g, '; however, the model requires at least one landunit per grid cell, so the model has put back in this grid cell the largest landunit.'`
`end if`

Currently 100% means NONE of this landunit will appear UNLESS the model needs it in an empty grid cell. You're suggesting that we add a > 100% option to ensure that a landunit never appears. I feel skeptical about pursuing this for a couple of reasons:
1) A grid cell in some dataset may really have just one landunit and it could be the same landunit that the user chose to eliminate.
2) The toosmall_* parameters make sense as 0-100%, while >100% makes them potentially unintuitive, confusing, and prone to error.
3) If a user really wants a particular landunit entirely eliminated (say wetland), they can set toosmall_wetland = 100, the other toosmall_* = 0, and n_dom_landunits = some small number. This way wetland will only appear if all else is 0.

What do you think?

Reply:
...removing
`if ( masterproc )`
from the warning above...